File: 0018 四数之和/main.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        eq = []
        i=0
        while(i<len(nums)-3):
            j=i+1
            while(j<len(nums)-2):
                li = j+1
                ri = len(nums)-1
                while li != ri:
                    if(nums[i]+nums[j]+nums[li]+nums[ri] == target):
                        eq.append([nums[i],nums[j],nums[li],nums[ri]])
                        while nums[ri]==nums[ri-1] and li!=ri-1:
                            ri-=1
                        ri-=1
                    elif(nums[i]+nums[j]+nums[li]+nums[ri] > target):
                        while nums[ri]==nums[ri-1] and li!=ri-1:
                            ri-=1
                        ri-=1
                    else:
                        while nums[li]==nums[li+1] and li+1!=ri:
                            li+=1
                        li+=1
                while j<=len(nums)-3 and nums[j]==nums[j+1]:
                    j+=1
                j+=1
            while i<=len(nums)-4 and nums[i]==nums[i+1]:
                i+=1
            i+=1
        return eq

File: 0018 四数之和/main.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        vector<vector<int>> res;
        if(nums.size()<4)
            return res;

        int li=0,ri=0;
        int num=0;
        sort(nums.begin(),nums.end());
        for(int i =0; i<nums.size()-3;i++){
            if(i>0 && nums[i]== nums[i-1])
                continue;
            if(nums[i] + nums[i+1] + nums[i+2] + nums[i+3] > target)
                continue;
            ri=nums.size()-1;
            if(nums[i] +nums[ri-2]+nums[ri-1] + nums[ri] <target)
                continue;
            
            for(int j =i+1;j < nums.size()-2;j++){
                if(j>i+1 && nums[j] == nums[j-1])
                    continue;
                
                li=j+1;
                ri=nums.size()-1;
                if(nums[i] + nums[j] + nums[li] + nums[li+1] > target)
                    continue;
                if(nums[i] +nums[j]+nums[ri-1] + nums[ri] <target)
                    continue;

                while(ri>li){
                    num = nums[i] +nums[j]+nums[li]+nums[ri];
                    if( num < target || (li>j+1 && nums[li] == nums[li-1]))
                        li++;
                    else if(num > target || (ri < nums.size() -1 && nums[ri]  == nums[ri+1]))
                        ri--;
                    else{
                        res.push_back({nums[i],nums[j],nums[li],nums[ri]});
                        li++;
                    }
                }
            }
        }
        return res;
    }
};